mm2src/mm2_main/src/lp_network.rs

```diff
@@ -138,36 +139,68 @@ async fn process_p2p_message(
     mut message: GossipsubMessage,
     i_am_relay: bool,
 ) {
+    fn is_valid(topics: &Vec<TopicHash>) -> bool {
+        if topics.is_empty() {
+            return false;
+        }
+
+        let first_topic_prefix = topics[0].as_str().split(TOPIC_SEPARATOR).next().unwrap_or_default();
+        for item in topics.iter().skip(1) {
+            if !item.as_str().starts_with(first_topic_prefix) {
+                return false;
+            }
+        }
+
+        true
+    }
+
     let mut to_propagate = false;
     let mut orderbook_pairs = vec![];
 
     message.topics.dedup();
     drop_mutability!(message);
 
+    if !is_valid(&message.topics) {
+        return;
+    }
+
     for topic in message.topics {
         let mut split = topic.as_str().split(TOPIC_SEPARATOR);
         match split.next() {
             Some(lp_ordermatch::ORDERBOOK_PREFIX) => {
                 if let Some(pair) = split.next() {
+                    if decode_signed::<OrdermatchMessage>(&message.data).is_err() {
+                        return;
+                    };
+
                     orderbook_pairs.push(pair.to_string());
                 }
             },
             Some(lp_swap::SWAP_PREFIX) => {
-                lp_swap::process_msg(ctx.clone(), split.next().unwrap_or_default(), &message.data).await;
+                if lp_swap::process_msg(ctx.clone(), split.next().unwrap_or_default(), &message.data)
+                    .await
+                    .is_err()
+                {
+                    return;
+                }
+
                 to_propagate = true;
+                break;
             },
             Some(lp_swap::WATCHER_PREFIX) => {
-                if ctx.is_watcher() {
-                    lp_swap::process_watcher_msg(ctx.clone(), &message.data).await;
+                if ctx.is_watcher() && lp_swap::process_watcher_msg(ctx.clone(), &message.data).is_err() {
+                    return;
                 }
+
                 to_propagate = true;
+                break;
             },
```

Comment:
don't you think you need to check for decode result in this block also?.. `process_msg` already does that but `process_watcher_msg`

```rust
Some(lp_swap::WATCHER_PREFIX) => {
                 if ctx.is_watcher() && lp_swap::process_watcher_msg(ctx.clone(), &message.data).is_err() {
                     return;
                 }

                 to_propagate = true;
                 break;
             },
``` 

like you in here

```rust
 Some(lp_ordermatch::ORDERBOOK_PREFIX) => {
                 if let Some(pair) = split.next() {
                     if decode_signed::<OrdermatchMessage>(&message.data).is_err() {
                         return;
                     };

                     orderbook_pairs.push(pair.to_string());
                 }
             },

```

Reply:
Orderbook is handled differently. It would collect all pairs of data without verifying if the messages were valid or not. That's why decoding check is added there.

As for the watcher, it already calls its function where it attempts to decode the message before proceeding.

https://github.com/KomodoPlatform/atomicDEX-API/blob/e520117a055f920b877ae701c679b9ffabc3b2ee/mm2src/mm2_main/src/lp_swap/swap_watcher.rs#L517-L524

Reply:
I see. Thanks for the clarification 🙏🏼